Event summary report WIP

andrew-1234 opened the pull request:
Draft pull request for audio_event_report

Fixes: #653

andrew-1234 committed:
adds new audio event report module, and associated controller, route, request spec

andrew-1234 committed:
Provisional report module structure under modules/report for organisation. outdated audio_event_report.rb in modules due to ongoing refactoring.

- Adds Report::Base class to serve as a foundation for report generation.
- Adds Report::AudioEventReport class for generating audio event reports with filtering.
- Adds Report::TimeSeries module for shared time series report logic.
- Changes to audio_event_factory
- Added audio_recording_factory with date range traits for report tests.
- Implemented specs for Report::AudioEventReport (WIP) and Report::TimeSeries .

- Implemented array_to_json method in Baw::Arel for JSON conversion of arrays.
- Created ArrayToJson node in Baw::Arel::Nodes for Arel SQL generation.

andrew-1234 committed:
WIP
- adds json_agg and row_to_json methods to Baw::Arel module + JsonAgg and RowToJson classes in Baw::Arel::Nodes for Arel query building.
- refactor the accumulation series result query to arel
- adds event summary results to report, sans score series
- implements an AudioEventReport formatter method; postgres decoding of the PG result object, and additional formatting/processing steps
- adds results processing methods as `trasnform_*` that map over result subsets

andrew-1234 committed:
Refactor WIP
- moved most raw sql into arel >> breaks time series spec (todo)
- refactored report sections into a basic cte creation pattern
- introduced  TimeSeries::Accumulation to put the accumulation series logic somewhere for now.

andrew-1234 committed:
reshuffle + document

andrew-1234 committed:
fix method call

andrew-1234 committed:
adds tag composition series to report output + formatting

andrew-1234 committed:
WIP

atruskie commented on app/modules/report/audio_event_report.rb:
```diff
@@ -0,0 +1,552 @@
+      event_summary_ctes, event_summaries_aggregate = event_summary_result(base_table)
+
+      # hack. access the table using .left
+      bucketed_time_series = accumulation_series_ctes.find { _1.left.name == 'bucketed_time_series' }
+      base_verification = event_summary_ctes.find { _1.left.name == 'verification_base' }
```
magic string

atruskie commented on app/modules/report/audio_event_report.rb:
```diff
@@ -0,0 +1,552 @@
+
+    # returns an expression that projects the absolute start time of an audio
+    # event as a derived columnn
+    def start_time_absolute_expression
+      'audio_recordings.recorded_date + CAST(audio_events.start_time_seconds || \' seconds\' as interval) ' \
+        'as start_time_absolute'
+    end
+
+    # returns an expression that projects the absolute end time of an audio
+    # event as a derived columnn
+    def end_time_absolute_expression
+      'audio_recordings.recorded_date + CAST(audio_events.end_time_seconds || \' seconds\' as interval) ' \
+        'as end_time_absolute'
+    end
```
pretty sure these already exist elsewhere in the solution

helpers like this can go on the AudioEvent model

atruskie commented on app/modules/report/time_series/coverage.rb:
```diff
@@ -0,0 +1,405 @@
+      # event as a derived columnn
+      def arel_recorded_end_date
+        # AudioRecording.arel_table[:recorded_date].+(Arel::Nodes.build_quoted('3600 seconds'))
+        Arel::Nodes::SqlLiteral.new('audio_recordings.recorded_date + CAST(audio_recordings.duration_seconds || \' seconds\' as interval)')
```
seems out of place here

andrew-1234 committed:
wip

andrew-1234 committed:
wip pre-refactor

andrew-1234 committed:
Refactor(report): adopt declarative Section/CTE pipeline & remove previous implementations

Introduce Report::Section DSL + Step class for modular, table-driven pipelines
Build out Section modules for each part of the report:
• Accumulation
• Composition
• Coverage
• EventSummary
Adds SQL matching specs for each report section
Adds TableExpression::Datum & Collection to manage CTE nodes & dependencies in topological order
Refactor AudioEventReport to drive its CTEs through the new Section pipeline and centralize JSON/array decoding in Ruby
Idea being to eventually deprecate the need for a base report class + specific report subclasses and move to pipeline only, with some kind of schema that can be defined on a report controller.
Remove outdated files/modules (Expression, Exp, old time_series/*, cte_definitions, temp helpers, etc.)
Updates request spec for /audio_event_reports to use the new shape
Removes experimental code, temp files, unused specs and helpers

andrew-1234 commented on app/controllers/audio_event_reports_controller.rb:
```diff
@@ -0,0 +1,32 @@
+    Access::ByPermission.audio_events(current_user)
+  end
+
+  def respond_report(content, opts = {})
+    content_type = 'application/json'
+    built_response = Settings.api_response.build(:ok, content)
+
+    if request.head?
+      head :ok, { content_length: built_response.to_json.bytesize, content_type: }
+    else
+      render json: built_response, status: :ok, content_type:, layout: false
+    end
+  end
```
fixed (I think?)  changed this to just:
 `render json: result, status: :ok`

andrew-1234 committed:
update report controller render response

andrew-1234 committed:
moves test files

atruskie commented on config/routes.rb:
```diff
@@ -781,6 +781,8 @@ def matches?(request)
       concerns: [:filterable]
   end
 
+  post '/audio_event_reports' => 'audio_event_reports#filter', defaults: { format: 'json' }
```
```suggestion
  post '/reports/audio_event_summary' => 'reports#audio_event_summary', defaults: { format: 'json' }
```

atruskie commented on spec/factories/audio_recording_factory.rb:
```diff
@@ -105,7 +105,19 @@
 
+    trait :with_date_range do
+      transient do
+        start_date { '2025-01-01 T00:00:00Z' }
```
if you just make this a date value you don't have to parse it on each sequence generation step

atruskie commented on spec/lib/modules/report/.time_series_spec.rb:
ignoring

atruskie commented on app/modules/report/audio_event_report.rb:
```diff
@@ -0,0 +1,259 @@
+      recording_coverage_options = Report::Section::Coverage.options(time_series_options) { |opt|
+        opt[:source] = base_table
+        opt[:lower_field] = base_table[:recorded_date]
+        opt[:upper_field] = Report::ArelHelpers.arel_recorded_end_date(base_table) # using AudioRecording.arel_recorded_end_date => missing FROM-clause entry for table "audio_recordings"
```
make the function on AudioRecording accept an optional arel_table argument. If not specified just defaults to `self.class.arel_table`

atruskie commented on app/modules/report/audio_event_report.rb:
```diff
@@ -0,0 +1,259 @@
+        opt[:source] = base_table
+        opt[:lower_field] = base_table[:recorded_date]
+        opt[:upper_field] = Report::ArelHelpers.arel_recorded_end_date(base_table) # using AudioRecording.arel_recorded_end_date => missing FROM-clause entry for table "audio_recordings"
+        opt[:analysis_result] = false
```
leaky

atruskie reviewed (commented):
Generally, unless you have calendar specific dates, use `Time` instead of `DateTime`

I've only had time for a partial review. I'll pick this up tomorrow.

atruskie commented on app/modules/report/section/coverage.rb:
```diff
@@ -0,0 +1,364 @@
+      TABLE_COVERED_DURATIONS = Arel::Table.new('covered_durations')
+      TABLE_FINAL_COVERAGE = Arel::Table.new('final_coverage')
+
+      TABLES = {
+        gap_size: TABLE_GAP_SIZE,
+        sort_with_lag: TABLE_SORT_WITH_LAG,
+        grouped: TABLE_GROUPED,
+        coverage_intervals: TABLE_COVERAGE_INTERVALS,
+        coverage_events: TABLE_COVERAGE_EVENTS,
+        coverage_events_sorted_sum: TABLE_COVERAGE_EVENTS_SORTED_SUM,
+        covered_intervals: TABLE_COVERED_INTERVALS,
+        covered_durations: TABLE_COVERED_DURATIONS,
+        final_coverage: TABLE_FINAL_COVERAGE
+      }.freeze
```
i've seen this structure a few times, seems a little redundant? what is it used for?

if you need a list of tables you could register them in your `step` method

atruskie commented on app/modules/report/section/step.rb:
i think the placement of this class is confusing. It should be next to section , rather than the sections

atruskie commented on app/modules/report/table_expression.rb:
```diff
@@ -0,0 +1,5 @@
+
+module Report
+  module TableExpression end
+end
```
redundant?

atruskie commented on app/modules/report/table_expression/collection.rb:
```diff
@@ -0,0 +1,116 @@
+    class Collection
+      def initialize(collection = {})
+        validate collection
+        @entries = collection.is_a?(Hash) ? collection.dup : collection.entries.dup
```
why not enforce one type of collection?

atruskie replied:
why duping? expensive if not needed

atruskie commented on app/modules/report/table_expression/collection.rb:
```diff
@@ -0,0 +1,116 @@
+      # @param keys [Array<Symbol>] The keys to subset the collection with
+      # @return [Report::Expression::Collection] New collection subset
+      def get(keys)
+        input = keys.is_a?(Array) ? keys.map(&:to_sym) : [keys.to_sym]
```
you can use the `Array(keys)` trick here

atruskie commented on app/modules/report/table_expression/datum.rb:
```diff
@@ -0,0 +1,31 @@
+
+module Report
+  module TableExpression
+    # An abstraction for Common Table Expressions (CTE).
+    #
+    # Combines an Arel::Table, Arel::Nodes::As CTE node, and optional
+    # dependency list for managing CTE relationships.
+    #
+    # @see Report::TableExpression::Collection to manage multiple Datum objects
+    class Datum
+      # @param table [Arel::Table] The CTE's table
+      # @param cte [Arel::Nodes::As] The CTE node
+      # @param dependencies [Array<Symbol>] Names of CTE dependencies
+      def initialize(table, cte, depends_on = [])
+        @table = table
+        @cte = cte
+        @depends_on = depends_on
+      end
+
+      # @return [Arel::Table] The Arel table for the CTE
+      attr_reader :table
+
+      # @return [Arel::Nodes::As] The CTE node
+      attr_reader :cte
+
+      # @return [Array<Symbol>] The names of CTE dependencies
+      attr_reader :depends_on
+    end
```
`Data.define` might be a good candidate for this class - if you keep it as is.

Though it could expand to be a lot more. Imagine if all your `step`s were just classes that inherited from this one?

`Datum` is not a great name for this structure. Maybe `CteNode`? Because nodes in a graph...

atruskie commented on spec/lib/modules/report/collection_spec.rb:
```diff
@@ -0,0 +1,38 @@
+
+describe 'TableExpression::Collection' do
+  # the correct order is A -> B -> C -> E -> D
+  test_data = [['table_a'], ['table_b'], ['table_c'], ['table_d'], ['table_e']]
+  dependencies = [[], [:table_a], [:table_b], [:table_a, :table_e], [:table_c]]
```
why not have a set of your `Datum` classes here?

atruskie commented on spec/lib/modules/report/collection_spec.rb:
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+describe 'TableExpression::Collection' do
+  # the correct order is A -> B -> C -> E -> D
```
From a documentation perspective, maybe show the DAG:

A -> B -> C -> E -> D
\\---------------------^

atruskie commented on spec/lib/modules/report/collection_spec.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  describe '#select' do
+    it 'returns a valid select statement for the specified key, with ordered CTEs' do
+      selected = collection.select(:table_c)
+      selected.to_sql
+      expected_sql = <<~SQL.squish
+        WITH "table_a" AS (SELECT), "table_b" AS (SELECT), "table_c" AS (SELECT) SELECT * FROM "table_c"
+      SQL
+      expect(selected.to_sql).to eq(expected_sql)
```
nice

atruskie reviewed (changes requested):
I think i was expecting your ctes/steps to be in different files. 

do have any common ctes between different sections?

andrew-1234 committed:
refactor the existing sql in audio_event model's  start/end scopes to class methods and use them in the audio event report

andrew-1234 committed:
Update app/controllers/audio_event_reports_controller.rb

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

andrew-1234 commented on app/modules/report/audio_event_report.rb:
```diff
@@ -0,0 +1,259 @@
+    end
+
+    def self.transform_tsrange(row)
+      matches = row['range'].match(/\["([^"]+)","([^"]+)"\)/)
```
using this worked thanks. the output now looks like:
`{:range=>"2025-01-03 00:00:00 UTC...2025-01-03 16:40:00 UTC",  .. }`

instead of an array before `['2025-01-02T00:00:00.000+00:00', '2025-01-03T00:00:00.000+00:00']`
is that good?

atruskie replied:
Yeah so, I think what's happening is we get a Ruby range, and the default JSON serializer is to_s/inspecting the value.

It's not great for two reasons: 1) the `...` synax is a distinctly Ruby idiom (AFAIK?) and 2) The date strings are a different format.

I'd either register a JSON serializer for the Range object to emit the tuple variant or do a post query transform on range objects. 

Or we could not use the postgres serializer.


---

Aside: using a two-values tuple for spans/intervals is in no way standard, but it is something I've seen used commonly in D3. I'm opting for that pattern because it is simple and space efficient in JSON, but it's arguably not self-describing in any way.

andrew-1234 committed:
start changes to generic reports naming
improve
review wip

andrew-1234 committed:
tests pass still, wip

atruskie commented:
> > I think i was expecting your ctes/steps to be in different files.
> > do have any common ctes between different sections?
> 
> yea there are at least a couple of CTEs that are defined in one section and required by a different section. would make sense to have these in their own files? in which case, should all ctes be in their own files? unless they are tightly coupled, then they could share a file?
> 
> That way a section wouldn't need to include all of the specific steps/CTE logic (like they do now), they could just define the main select statement needed to get a certain result (e.g. `project.composition.array_agg.as('composition')`, and the collection of CTEs required to execute it.
> 
> And then a concrete report implementation is just a template of which sections to include.
> 
> CTEs -> section = Collection of CTEs -> report = collection of sections

all in their own files? it's a bit overkill but for anything remotely complex it's more maintainable. I would, I think.

the rest sounds good!

andrew-1234 committed:
WIP WIP WIP
See README_cte_dsl.rb for the updated design. The new Cte module in that file,  is essentially two things: a class LazyCteNode, and a DSL for defining CTEs, will replace the need for the existing classes/modules like steps, step dsl, sections, nodes, collections.

andrew-1234 committed:
fix syntax in example
refactor in progress

andrew-1234 committed:
refactor cte.rb into cte module subdirectory (report/cte/**.rb)

add tests for  Report::Cte::Node (report/cte/node_spec.rb temporary file location WIP)

andrew-1234 committed:
wip

andrew-1234 committed:
wip converting to new cte node dsl
change eval method of select context
changes dsl to allow cte templating with explicit node subclassing + dsl, as well as original inline  style declarations

andrew-1234 committed:
WIP accumulation refactor + tests

andrew-1234 committed:
adds comparison_sql to rubocop config to fix NoExpectationExample offences
wip refactor continued with updated event composition and accumulation ctes + tests

andrew-1234 committed:
wip

andrew-1234 committed:
all ctes updated to new syntax, full clean-up + removal of old logic/files  is needed next , and fix namespacing / organisation
representing reports themselves as a cte node that depend on cte nodes is working

andrew-1234 committed:
cte refactoring

andrew-1234 committed:
wip, adding decoders, fixing base event class default scope

andrew-1234 committed:
add arel attr distinct helper
adds decode module,
adds formatters + result decoders to ctes

andrew-1234 committed:
add formatter to event summary cte

andrew-1234 committed:
the original request/ spec for the report is once again  passing,  using the refactored report and ctes code

andrew-1234 committed:
adds node inspect
adds #new_with_suffix  to node

andrew-1234 committed:
fix bug typo in audio recording model scope

andrew-1234 commented on app/modules/report/arel_helpers.rb:
```diff
@@ -0,0 +1,22 @@
+    # @param table [Arel::Table] The table to aggregate from
+    # @param field [Symbol] The field to aggregate
+    # @return [Arel::Nodes::NamedFunction] An aggregation node
+    def aggregate_distinct(table, field)
+      Arel::Nodes::NamedFunction.new(
+        'ARRAY_AGG',
+        [Arel::Nodes::SqlLiteral.new("DISTINCT #{table[field].name}")]
+      )
+    end
```
nice, added the distinct helper on attribute and now can compose like:
`base_table[:site_id].distinct.array_agg.as('site_ids')`

andrew-1234 committed:
removes unneeded arel helpers file
update audio_event model's arel_start/end helpers to return arel not string

andrew-1234 committed:
reverts api/response.rb

andrew-1234 commented on app/modules/report/section/composition.rb:
```diff
@@ -0,0 +1,137 @@
+        composition_series_aliased = TABLE_COMPOSITION_SERIES.as('c')
+        Arel::SelectManager.new
+          .project(composition_series_aliased.right.json_agg)
+          .from(composition_series_aliased)
```
nah, the 'composition_series_aliased' is an arel::nodes::as so you get undefined method `project` 
there's probably a better way than this weird pattern but i'm not sure how

andrew-1234 committed:
adds some validation. experimental

andrew-1234 committed:
refactor

andrew-1234 committed:
fixes issues from refactor
requests/audio_event_reports_spec.rb passing again

andrew-1234 committed:
fix incorrect method name

andrew-1234 committed:
improve error handling for dependency initializing
add dependency initializer spec
move related ctes into subfolder

andrew-1234 committed:
documentation

andrew-1234 committed:
fix error in event summary score histogram, incorrectly grouped
fix the raw result comparison test  that wasn't matching exactly
adds documentation

atruskie commented:
We went with a different architecture after doing real world performance benchmarks on production data. Still this work was an invaluable first iteration, thank you.
